### ISSUE TITLE
"derive makeArbitrary" error w/ strict types (8.0.1)

### ISSUE THREAD

**Issue by jonschoning** [ndmitchell/derive#17]
```
data NoteListItem = NoteListItem {
      noteListItemId     :: !Text
    , noteListItemHash   :: !Text
    , noteListItemTitle  :: !Text
    , noteListItemLength :: !Int
    , noteListItemCreatedAt :: !UTCTime
    , noteListItemUpdatedAt :: !UTCTime
    } deriving (Show, Eq, Data, Typeable, Generic, Ord)


$(derive makeArbitrary ''NoteListItem)


Exception when trying to run compile-time code:
      Could not convert (Bang,Type) to Type
(Bang NoSourceUnpackedness SourceLazy,ConT Data.Text.Internal.Text)
src/Language/Haskell/Convert.hs:(76,5)-(79,64): Non-exhaustive patterns in function conv

CallStack (from HasCallStack):
  error, called at src/Language/Haskell/Convert.hs:20:87 in derive-2.5.25-6jyS0d772bZHapQdVyiucF:Language.Haskell.Convert
    Code: derive makeArbitrary ''NoteListItem
```

**Comment by ndmitchell** [comment]
Works for me using Derive 2.5.25 and the test case below. Can you check with that test case and Derive version, and we can see where things differ.

```haskell
{-# LANGUAGE TemplateHaskell, DeriveDataTypeable, DeriveGeneric, StandaloneDeriving #-}

module Example where

import Data.Text
import Data.Data
import GHC.Generics
import Data.Time.Clock
import Data.Derive.Arbitrary
import Test.QuickCheck
import Data.DeriveTH
import System.IO.Unsafe

data NoteListItem = NoteListItem {
      noteListItemId     :: !Text
    , noteListItemHash   :: !Text
    , noteListItemTitle  :: !Text
    , noteListItemLength :: !Int
    , noteListItemCreatedAt :: !UTCTime
    , noteListItemUpdatedAt :: !UTCTime
    } deriving (Show, Eq, Data, Typeable, Generic, Ord)

instance Arbitrary Text where
    arbitrary = return mempty

instance Arbitrary UTCTime where
    arbitrary = return $ unsafePerformIO getCurrentTime

$(derive makeArbitrary ''NoteListItem)


main = print =<< generate (arbitrary :: Gen NoteListItem)
```

**Comment by jonschoning** [comment]
The test case you posted above works fine - no issues.

I've created a minimal repo that reproduces the failing behavior here: https://github.com/jonschoning/derive-issue-17

**Comment by jonschoning** [comment]
another side note - what I don't quite understand is where it is getting `SourceLazy` from in `Bang NoSourceUnpackedness SourceLazy`... I would think it should interpret all the fields as `SourceStrict` due to the bangs.
```
>>> import Derive17 
>>> :kind! Rep NoteListItem 
Rep NoteListItem :: * -> *
= D1
    ('MetaData
       "NoteListItem"
       "Derive17"
       "derive17-0.0.1-G2Y3UnpjaWJD80ZsMDbSig"
       'False)
    (C1
       ('MetaCons "NoteListItem" 'PrefixI 'True)
       ((S1
           ('MetaSel
              ('Just "noteListItemId")
              'NoSourceUnpackedness
              'SourceStrict
              'DecidedStrict)
           (Rec0 TE.Text)
         :*: (S1
                ('MetaSel
                   ('Just "noteListItemHash")
                   'NoSourceUnpackedness
                   'SourceStrict
                   'DecidedStrict)
                (Rec0 TE.Text)
              :*: S1
                    ('MetaSel
                       ('Just "noteListItemTitle")
                       'NoSourceUnpackedness
                       'SourceStrict
                       'DecidedStrict)
                    (Rec0 TE.Text)))
        :*: (S1
               ('MetaSel
                  ('Just "noteListItemLength")
                  'NoSourceUnpackedness
                  'SourceStrict
                  'DecidedUnpack)
               (Rec0 Int)
             :*: (S1
                    ('MetaSel
                       ('Just "noteListItemCreatedAt")
                       'NoSourceUnpackedness
                       'SourceStrict
                       'DecidedStrict)
                    (Rec0 UTCTime)
                  :*: S1
                        ('MetaSel
                           ('Just "noteListItemUpdatedAt")
                           'NoSourceUnpackedness
                           'SourceStrict
                           'DecidedStrict)
                        (Rec0 UTCTime)))))
```

**Comment by ndmitchell** [comment]
Thanks for the awesome test case - I've got a fix, just doing a buildbot test and then I'll release it.

I suspect the constructors are either strict, lazy or default. Within a module they are default, but once they leave a module they get given explicit values.

**Comment by jonschoning** [comment]
oh ok.. thank you!  I will read up on this behavior

**Comment by ndmitchell** [comment]
That behaviour is a complete guess - but it seems to fit with what we're seeing. I expect the two different modules is important.

**Comment by ndmitchell** [comment]
Fixed in 2.5.26. Thanks for the awesome test case!
